pyrsistent/_pmap.py

```diff
@@ -125,7 +125,27 @@ def __len__(self):
     def __repr__(self):
         return 'pmap({0})'.format(str(dict(self)))
 
-    __eq__ = Mapping.__eq__
+    def __eq__(self, other):
+        if self is other:
+            return True
+        if not isinstance(other, Mapping):
+            return NotImplemented
+        if len(self) != len(other):
+            return False
+        if isinstance(other, PMap):
+            try:
+                if hash(self) != hash(other):
+                    return False
+            except TypeError:
+                # self or other contains non-hashable values
+                pass
+            if self._buckets == other._buckets:
+                return True
+            return dict(self.iteritems()) == dict(other.iteritems())
```

Comment:
Yes, it's a vector, to achieve structural sharing. Currently it's O(n) but there's really no reason that a length comparison could not be applied as a first check for the vector as well to rule out a large number of cases of inequality.